feat: add snapping endpoint

sfendrich opened the pull request:
### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box
when you have checked you have done them): -->
- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [x] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [x] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes #1519 .

### Information about the changes
- Key functionality added: snapping endpoint that returns locations snapped to the nearest edge in the graph

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

Commit:
feat: add snapping endpoint

This commit adds a new service and endpoints that returns given
locations snapped to the nearest edge in the graph.

Commit:
docs: add changelog entry for snapping service

MichaelsJP commented on ors-api/src/main/java/org/heigit/ors/api/responses/snapping/SnappingResponse.java:
```diff
@@ -0,0 +1,12 @@
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import org.heigit.ors.snapping.SnappingResult;
+public class SnappingResponse {
+    @JsonIgnore
+    private final SnappingResult result;
+
+    public SnappingResponse(SnappingResult result) {
+        this.result = result;
+    }
```
Is this already sufficient, in terms of swagger docs?

TheGreatRefrigerator replied:
Yes, the Response object types are built from the locations array and JSONLocation Object
![grafik](https://github.com/GIScience/openrouteservice/assets/23240110/bed4fd3b-5fec-4c3d-b974-86bcfca2dcac)

MichaelsJP commented on ors-api/src/main/java/org/heigit/ors/api/responses/snapping/json/JsonSnappingResponse.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Schema(name = "SnappingResponse", description = "The Snapping Response contains the snapped coordinates.")
```
Even though the response is quite simple it could make sense to have response examples for the docs api

TheGreatRefrigerator replied:
I can see one example being generated in the swagger ui which is probably derived from the JSONLocation type:
![grafik](https://github.com/GIScience/openrouteservice/assets/23240110/a9805f58-798f-4a0e-b766-b2c16347f6c9)
This is ok i think.

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/routing/APIEnums.java:
```diff
@@ -172,6 +172,32 @@ public String toString() {
         }
     }
 
+    @Schema(name = "Snapping response type", description = "Format of the snapping response.")
```
@TheGreatRefrigerator Does the schema need more definition?

TheGreatRefrigerator replied:
If the enum is not used in the Body, this doesn't seem to be used for generating the OpenAPI specs. Same for the other enums in the file (RouteResponseType, MatrixResponseType, etc.).

So this is fine as is.

MichaelsJP commented on ors-api/src/test/java/org/heigit/ors/apitests/snapping/ParamsTest.java:
```diff
@@ -0,0 +1,48 @@
+                .statusCode(200);
+
+    }
+}
```
Imo a few more tests could be good.
At least one test checking failing requests.
Maybe two parametrized tests with different search radii could be better suited.

TheGreatRefrigerator replied:
agree

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/snapping/SnappingRequest.java:
```diff
@@ -0,0 +1,49 @@
+        String CHProfileName = ProfileTools.makeProfileName(encoderName, hintsMap.getString("weighting", ""), false);
+        String profileName = CHProfileName;
+
+        // TODO: replace usage of matrix search context by snapping-specific class
```
What does the to do depend on, could this already be part of this PR?

sfendrich replied:
It mainly depends on priorities. Implementing these snapping-specific classes is only worth, if we do it cleanly. This would require significant effort with moderate impact. The list of more important cleanup issues is long.

aoles replied:
At the time being the snap endpoint **works only if CH has been enabled** for a given profile and fails otherwise 🙈 This happens because of the dependency on the `routingCHGraph` in the following call.

https://github.com/GIScience/openrouteservice/blob/b392a0c5681fe73afa1ae1559bb6452598fba772/ors-engine/src/main/java/org/heigit/ors/snapping/SnappingRequest.java#L41

TheGreatRefrigerator commented:
I can't run the normal Application run configuration of spring-boot.
It doesn't work with my old ors-config.json but even if i remove that and it uses the default files (after uncommenting the pbf line).
```
11 Aug 16:12:13 INFO                                                  main [ o.h.o.a.Application                      ]   Starting Application using Java 17.0.6 with PID 26175 (/Users/abutzer/Workspace/openrouteservice/ors-api/target/classes started by abutzer in /Users/abutzer/Workspace/openrouteservice)
11 Aug 16:12:13 INFO                                                  main [ o.h.o.a.Application                      ]   The following 1 profile is active: "default"
11 Aug 16:12:15 INFO                                              ORS-Init [ o.h.o.a.s.l.ORSInitContextListener       ]   Initializing ORS...
11 Aug 16:12:15 INFO                                              ORS-Init [ o.h.o.r.RoutingProfileManager            ]   Total - 110 MB, Free - 82.99 MB, Max: 4 GB, Used - 27.01 MB
11 Aug 16:12:15 ERROR                                             ORS-Init [ o.h.o.r.RoutingProfileManager            ]   
11 Aug 16:12:15 ERROR                                             ORS-Init [ o.h.o.r.RoutingProfileManager            ]   No profiles configured. Exiting.
11 Aug 16:12:15 ERROR                                             ORS-Init [ o.h.o.r.RoutingProfileManager            ]   
11 Aug 16:12:15 INFO                                                  main [ o.h.o.a.Application                      ]   Started Application in 2.803 seconds (process running for 3.698)
11 Aug 16:12:15 INFO                         SpringApplicationShutdownHook [ o.h.o.a.s.l.ORSInitContextListener       ]   Shutting down ORS and releasing resources.

```

TheGreatRefrigerator commented on ors-api/src/main/java/org/heigit/ors/api/controllers/SnappingAPI.java:
```diff
@@ -0,0 +1,157 @@
+    @Operation(
+            description = """
+                    Returns a list of points snapped to the nearest edge in the graph. \
+                    This method does not accept any request body or parameters other than profile, locations, maximum_snapping_radius.\
```
Is this line neccessary? Which parameters the endpoint accepts and the type is obvious with the swagger ui / API playground.

Probably replace with some info on Locations not found which result in `null`.

sfendrich replied:
Done.

TheGreatRefrigerator commented on ors-api/src/main/java/org/heigit/ors/api/controllers/SnappingAPI.java:
```diff
@@ -0,0 +1,157 @@
+
+    @PostMapping(value = "/{profile}/json", produces = {"application/json;charset=UTF-8"})
+    @Operation(
+            description = "Returns a list of points snapped to the nearest edge in the graph.",
```
This should probably also contain the description of the other snapping endpoint, as they do the same.

Is there a reason for having always the `/json` endpoint as well?

sfendrich replied:
Done.

TheGreatRefrigerator replied:
the `/json` is there as we have the same structure in the other apis as well. Once we have `/geojson` it makes more sense again as well.

TheGreatRefrigerator commented on ors-api/src/main/java/org/heigit/ors/api/controllers/SnappingAPI.java:
```diff
@@ -0,0 +1,157 @@
+            responseCode = "200",
+            description = "Standard response for successfully processed requests. Returns JSON.",
+            content = {@Content(
+                    mediaType = "application/geo+json",
```
The response is not a geojson so this should be `application/json`. 
You can validate geojson with:
http://geojson.io/

See e.g. https://en.wikipedia.org/wiki/GeoJSON for geojson format.

Should we have that as response format as well?

sfendrich replied:
Done.

TheGreatRefrigerator commented on ors-api/src/main/java/org/heigit/ors/api/controllers/SnappingAPI.java:
```diff
@@ -0,0 +1,157 @@
+            responseCode = "200",
+            description = "JSON Response.",
+            content = {@Content(
+                    mediaType = "application/geo+json",
```
see above

sfendrich replied:
Done.

TheGreatRefrigerator reviewed (changes requested):
One general question we should consider is wether we want to return this as geojson as well?

apart from that some smaller issues to resolve.

TheGreatRefrigerator commented:
@MichaelsJP @sfendrich Do we want this endpoint on the live API or should it be hidden in the Playground?
If yes we also need to consider this endpoint in the tyk policies and quota plans.

MichaelsJP commented:
> @MichaelsJP @sfendrich Do we want this endpoint on the live API or should it be hidden in the Playground? If yes we also need to consider this endpoint in the tyk policies and quota plans.

I'd say we open it up on the live API with the next major Release 8.

MichaelsJP commented:
> One general question we should consider is wether we want to return this as geojson as well?
> 
> apart from that some smaller issues to resolve.

@TheGreatRefrigerator Should be effortlessly integrable and would make the endpoint similar to the others. I'd say let's do it.

MichaelsJP commented:
So, just three things missing:
1. Decide on a request limit. We need one. If it's possible, we could just do it in the API-Gateway. It would make our lives easier.
2. Add a few more test cases (parametrized) and failing.
3. Add the GeoJSON return.
4. Test performance

Adjust the documentation?

MichaelsJP commented:
> So, just three things missing:
> 
> 1. Decide on a request limit. We need one. If it's possible, we could just do it in the API-Gateway. It would make our lives easier.
We should integrate a config level request limit for the number of points to query. Even though we might add another check in the future in the api gateway

MichaelsJP committed:
test(snapping): Add parametrized exception test

Co-authored-by: Jochen Haeussler <jochen.haeussler@heigit.org>

aoles reviewed (changes requested):
1. Add endpoint to `services` listed at localhost:8082/ors/v2/status.
2. Instead of the somewhat lengthy API parameter name `maximum_search_radius` wouldn't just `radius` be sufficient?

TheGreatRefrigerator committed:
refactor: extract EngineInfo class to common

EngineInfo was a subclass of:
- IsochronesResponseInfo
- MatrixResponseInfo
- RouteResponseInfo

It is now a common class for all responses

TheGreatRefrigerator committed:
feat: expose graphDate in SnappingResult

TheGreatRefrigerator committed:
feat: extend JsonSnappingResponse with metadata object

as in our other services, the snapping endpoint is now also returning
the metadata object containing info about the service and request made.

- add SnappingResponseInfo class
- extend SnappingAPI class with endpointsProperties and systemMessageProperties
  needed for generic response metadata

MichaelsJP committed:
test(snapping): refactor basicTest to testSnappingSuccess

- use argument stream for parametrization
- add multiple test cases

MichaelsJP committed:
feat(status): Add snap service to config and status page

MichaelsJP commented:
Snap is added to Status

TheGreatRefrigerator committed:
feat: add Getter for name of matrix JSONLocation schema

jhaeu committed:
test(snapping): add endpoint and profile conditionally

- add more tests with missing endpoint/profile

aoles reviewed (dismissed):
~The following TODO  needs to be addressed.~

https://github.com/GIScience/openrouteservice/blob/b392a0c5681fe73afa1ae1559bb6452598fba772/ors-engine/src/main/java/org/heigit/ors/snapping/SnappingRequest.java#L38

aoles committed:
refactor(snapping): remove dependency on CH graph

jhaeu committed:
refactor(snapping): moved computing method to the service class

TheGreatRefrigerator committed:
feat(snapping): add GeoJSON endpoint

- add schema classes for GeoJSON response
- add endpoint to SnappingAPI
- add GeoJSON to APIEnums

TheGreatRefrigerator committed:
docs: add unit to snapped_distance of JSONLocation schema

jhaeu committed:
test(snapping): refactored ParamsTest and covered geojson

jhaeu committed:
feat(snapping): add source_id to geojson response

TheGreatRefrigerator committed:
docs: add info for snap service to config docs

TheGreatRefrigerator committed:
docs: add attribution description & fix table

TheGreatRefrigerator reviewed (approved):
looks ready now
